File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        if(nums.size()==1) return nums[0]==target?0:-1;
        int mid, l = 0, r = nums.size() - 1;
        while (l < r) {
            mid = l + (r - l) / 2;
            if (nums[mid] > nums[r])
                l = mid + 1;
            else
                r = mid;
        }
        int pivot = l,n=nums.size();
        l = 0, r = nums.size() - 1;
        // cout<<pivot<<endl;
        while(l<=r){
            mid=l+(r-l)/2;
            int realidx = (pivot+mid)%n;
            // cout<<realidx<<endl;
            if(nums[realidx]==target) return realidx;
            if(nums[realidx]>target) r=mid-1;
            else l=mid+1;
        }
        return nums[l]==target?l:-1;
    }
};

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:

    def binsearch(self,nums,l,r,target): 
        while l<=r: 
            mid = l+(r-l)//2 
            if nums[mid]==target: 
                return mid 
            elif nums[mid]>target: 
                r=mid-1 
            else: 
                l=mid+1 
        return -1 

    def search(self, nums: List[int], target: int) -> int:
        l,r = 0,len(nums)-1 

        peak=-1 
        while l<=r: 
            mid = l+(r-l)//2 
            if nums[mid]>nums[-1]: 
                peak=mid
                l=mid+1 
            else: 
                r=mid-1 

        if peak==-1: 
            return self.binsearch(nums,0,len(nums)-1,target) 
        if nums[0]<=target<=nums[peak]:
            return self.binsearch(nums,0,peak,target) 
        return self.binsearch(nums,peak+1,len(nums)-1,target)